src/Microsoft.OData.Core/ODataLibraryCompatibility.cs

```diff
@@ -4,15 +4,15 @@
 // </copyright>
 //---------------------------------------------------------------------
 
-using System.Diagnostics.CodeAnalysis;
+using System;
 
 namespace Microsoft.OData
 {
     /// <summary>
     /// Library compatibility levels.
     /// </summary>
-    [SuppressMessage("Microsoft.Design", "CA1008:EnumsShouldHaveZeroValue")]
-    public enum ODataLibraryCompatibility
+    [Flags]
+    public enum ODataLibraryCompatibility : long
     {
         /// <summary>
         /// Version 6.x

```

Comment:
@mikepizzo suggestion was to include all the options that correspond to Version6 and Version7 as individual flags, and the Version6 and Version7 would be a combination of all the flags that correspond to v6 and v7 compatibility respectively. It doesn't seem like that was implemented. Do you plan to have that done in a separate PR (I think that's okay, but then it should be tracked explicitly).

If we're going to keep `Version6` and `Version7` as they are (i.e. not a combination of flags), we should consider that we may have many flags for all the different compatibility settings we may introduce. And so we should consider moving the `06` and `07` to move significant bits i.e. add more 0s to the right. Not that now that you've changed the type to long, we can have up to 64 independent flags. If we use one octal digit for the version (assuming the version is not a combination of flags), then that means only 61 digits  can be used as flags, but 7 is the largest digit we can represent with a single octal (3 bits), so with this approach we'd need 4 bits at least to represent version 8-15, which means fewer bits for the flags, which would be a problem if those bits were already used for some flags. As you can see this is not very sustainable or future proof.

Reply:
I don't think we currently have specific flags for v6 ad v7.. If we do. I'd like to have that in a separate PR:

Reply:
Is there anywhere in the library where ODataLibraryCompatibility.V6 or ODataLibraryCompatibility.V7 are referenced?

Reply:
Yes there are there.

Reply:
This is an example: https://github.com/OData/odata.net/blob/24ae2e335650c2533e541e1956f5b64f618ee0cd/src/Microsoft.OData.Core/JsonLight/ODataJsonLightPropertySerializer.cs#L649

Reply:
@ElizabethOkerio This is a good point that @habbes raised here.
I had forgotten some of the details around that discussion.
The idea was, e.g. for the line of code you have highlighted, we'd introduce a flag in the `ODataLibraryCompability` enum for the particular behaviour being controlled here.
As an illustration:
```csharp
public enum ODataLibraryCompatibility
{
    // ...
    SupportWritingTopLevelNullProperty = 1 << 1
    // ...

    Version6 = SupportWritingTopLevelNullProperty | AnotherVersion6BehaviourFlag | YetAnotherVersion6BehaviourFlag
}
```
Then, we'd modify the logic here to:
```csharp
                if ((this.JsonLightOutputContext.MessageWriterSettings.LibraryCompatibility & ODataLibraryCompatibility.SupportWritingTopLevelNullProperty) == ODataLibraryCompatibility.SupportWritingTopLevelNullProperty && this.JsonLightOutputContext.MessageWriterSettings.Version < ODataVersion.V401)
                {
                    // The 6.x library used an OData 3.0 protocol element in this case: @odata.null=true
                    this.ODataAnnotationWriter.WriteInstanceAnnotationName(ODataAnnotationNames.ODataNull);
                    this.JsonWriter.WriteValue(true);
                }
                else
                {
                    // From the spec:
                    // 11.2.3 Requesting Individual Properties
                    // ...
                    // If the property is single-valued and has the null value, the service responds with 204 No Content.
                    // ...
                    throw new ODataException(Strings.ODataMessageWriter_CannotWriteTopLevelNull);
                }
```

That way, the person who had code like the following would not need to change a thing:
```csharp
var messageWriterSettings = new ODataMessageWriterSettings
{
    // ...
    LibraryCompatibility = ODataLibraryCompatibility.Version6
};

// Inject messageWriterSettings into the DI container
```

Since `Version6` is a combination of different flags, the above code has the effect of turning on all flags that control writing a `Version6` payload.

This should also settle the debate we had here https://github.com/OData/odata.net/pull/2905/files#r1574367462

`Version6` and `Version7` flags should be assigned a combination of flags, not a specific value.

Reply:
To clarify the idea, we would have:
```csharp
Version6 = SupportWritingTopLevelNullProperty | UseLegacyVariableCasing | AnotherBackwardCompatiblityFlagIntroducedInODL8 | Etc
Version7 = UseLegacyVariableCasing | AnotherBackwardCompatiblityFlagIntroducedInODL8 | Etc
```
These flags come in very handy since someone doesn't need to toggle each individual flag. If they need fine control, then they can set individual flags.

Reply:
Those are the places I'm concerned about. The thing is, by making it a flags enum, you have effectively changed the semantics of the enum and how it should be used. I'm concerned that existing usages may no longer have thei intended meaning.

For example, the usage you linked `this.JsonLightOutputContext.MessageWriterSettings.LibraryCompatibility < ODataLibraryCompatibility.Version7` This assumes the enum values are ordered. But now that they are flags, we shouldn't assume an inherent order. This particular case might work because Version7's enum sets the most significant bit of all the other values currently in the enum. But I can see this breaking easily. For example, if we had something like `> ODataLibraryCompatibility.Version6` with the intention of targeting version7 and later, this would return true for scenarios where we combined Version6 with another flag like `Version6 | LegacyVariableCasing` which is not necessarily v7 but is numerically greater than `Version6`'s current enum representation.

Before we merge the PR, I think we should audit all existing references of Version6 and Version7 and ensure that moving ahead with the PR as-is (while waiting for follow-up PR) is a safe thing to do. Ultimately, I don't think this feature should be shipped in the final release (or release candidate) without having refactored existing usage of the enum.

For example, the scenario that you linked to is about top level nulls in OData V3 or something like that. Instead of using < Version7, it would be safer to create an enum value called something like `AllowTopLevelNullForODataV3` then test for that: `LibraryCompatibility & AllowTopLevelNullForODataV3 == AllowTopLevelNullForODataV3` and then `Version6` would redefined as combination of all enum values that are valid for Version6. But using `>` or `<` against these enums is now a red flag that should be carefully audited (and avoided when possible).

Reply:
@gathogojr  @habbes  I think I'm ok with changing the way we are checking for the specific enums for this PR but having to create specific enums for the various changes can be done in a different PR.

Reply:
I don't even think it's necessary for the v6 and v7 flags to be separate since the people who've been using this have been using just one flag..Maybe for v8 then we can have the individual flags and one flag that has everything.